power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -651,106 +710,64 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
         return std::ranges::all_of(update_independent, [](bool const is_independent) { return is_independent; });
     }
 
-    template <symmetry_tag sym> auto calculate_power_flow(Options const& options) {
+    // Single calculation
+    template <calculation_type_tag calculation_type, symmetry_tag sym> auto calculate(Options const& options) {
+        auto const calculator = [this, &options] {
+            if constexpr (std::derived_from<calculation_type, power_flow_t>) {
+                return calculate_power_flow_<sym>(options.err_tol, options.max_iter);
+            }
+            assert(options.optimizer_type == OptimizerType::no_optimization);
+            if constexpr (std::derived_from<calculation_type, state_estimation_t>) {
+                return calculate_state_estimation_<sym>(options.err_tol, options.max_iter);
+            }
+            if constexpr (std::derived_from<calculation_type, short_circuit_t>) {
+                return calculate_short_circuit_<sym>(options.short_circuit_voltage_scaling);
+            }
+            throw UnreachableHit{"MainModelImpl::calculate", "Unknown calculation type"};
+        }();
+
         return optimizer::get_optimizer<MainModelState, ConstDataset>(
-                   options.optimizer_type, options.optimizer_strategy,
-                   calculate_power_flow_<sym>(options.err_tol, options.max_iter),
+                   options.optimizer_type, options.optimizer_strategy, calculator,
                    [this](ConstDataset update_data) { this->update_component<permanent_update_t>(update_data); },
                    *meta_data_)
             ->optimize(state_, options.calculation_method);
     }
 
-    // Single load flow calculation, propagating the results to result_data
-    template <symmetry_tag sym>
-    void calculate_power_flow(Options const& options, MutableDataset const& result_data, Idx pos = 0) {
+    // Single calculation, propagating the results to result_data
+    void calculate(Options options, MutableDataset const& result_data, Idx pos = 0) {
         assert(construction_complete_);
-        auto const math_output = calculate_power_flow<sym>(options);
 
-        if (pos != ignore_output) {
-            output_result(math_output, result_data, pos);
-        }
-    }
+        if (options.calculation_type == CalculationType::short_circuit) {
+            auto const faults = state_.components.template citer<Fault>();
+            auto const is_three_phase = std::ranges::all_of(
+                faults, [](Fault const& fault) { return fault.get_fault_type() == FaultType::three_phase; });
+            options.calculation_symmetry =
+                is_three_phase ? CalculationSymmetry::symmetric : CalculationSymmetry::asymmetric;
+        };
 
```

Comment:
I know this is implementation details, but since it is only relevant for short circuit calculations, I think it should be moved to `short_circuit_calculation`, maybe within its own function. Also, the logic would be more straightforward, as in the other two types of calculations the symmetry is obtained there.

Reply:
The problem is that this was part of its own function before, but the logic would become convoluted. The whole point of this PR is to isolate the differences between the calculation types etc. in one place. In the new state after this PR, it is **very** clear that specifically short circuit calculations ignore the provided calculation symmetry. Earlier, that would be hidden behind the fact that some functions are templated and some are not.

Reply:
Understood. Now it is indeed very straight forward when one sees this.